clients/src/main/java/org/apache/kafka/common/utils/CircularIterator.java

```diff
@@ -14,36 +14,69 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+
 package org.apache.kafka.common.utils;
 
+import java.util.Collection;
 import java.util.Iterator;
-import java.util.List;
+import java.util.Objects;
 
+/**
+ * An iterator that cycles through the Iterator of a Collection indefinitely.
+ * Useful for tasks such as round-robin load balancing.
+ */
 public class CircularIterator<T> implements Iterator<T> {
-    int i = 0;
-    private List<T> list;
 
-    public CircularIterator(List<T> list) {
-        if (list.isEmpty()) {
+    private final Iterable<T> iterable;
+    private Iterator<T> iterator;
+    private T peek;
+
+    /**
+     * Create a new instance of a CircularIterator. The ordering of this
+     * Iterator will be dictated by the Iterator returned by Collection itself.
+     *
+     * @param col The collection to iterate indefinitely
+     *
+     * @throws NullPointerException if col is {@code null}
+     * @throws IllegalArgumentException if col is empty.
+     */
+    public CircularIterator(final Collection<T> col) {
+        this.iterable = Objects.requireNonNull(col);
+        this.iterator = col.iterator();
+        this.peek = null;
+        if (col.isEmpty()) {
             throw new IllegalArgumentException("CircularIterator can only be used on non-empty lists");
         }
-        this.list = list;
     }
 
     @Override
     public boolean hasNext() {
-        return true;
+        if (this.peek != null) {
```

Comment:
Why change the logic since the answer is always `true`?

Reply:
Astute observation.  So, technically speaking, it is possible to not return true.  Imagine the following code....

```
List a = new ArrayList<>(Arrays.asList("A"));
Iterator it = new CircularIterator(a);
System.out.println(it.next());
a.clear();
System.out.println(it.next());
```

The Iterator is re-created on the second call to `next()` so this is a legal option.  Now, not sure it would actually happen, but it's possible.

Reply:
The constructor doesn't accept an empty list, but it doesn't copy it (or, better, if we were using persistent data structures... but alas no).  The class isn't thread-safe, and iterators that the code constructs and uses are potentially fail-safe or fail-fast if the collection is modified.  And the `remove()` method of this class throws `UnsupportedOperationException`.

All in all, I think the best thing is to clearly document that the class does not support modifications to the passed-in collection and always return `true` here (technically !isEmpty() covers all cases, but since the constructor doesn't accept an empty list the answer would always be `true`).